File: Insulator/app/src/main/java/com/bendywalker/insulator/Card.java
package com.bendywalker.insulator;

import android.content.Context;
import android.content.res.TypedArray;
import android.text.Editable;
import android.text.Selection;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.RelativeLayout;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;


public class Card extends RelativeLayout {

    static boolean textChangeRunning;
    OnTextChangeListener textChangeListener;
    private TextView label, info;
    private EditText entry;
    private String prefKey;
    private MyPreferenceManager preferenceManager;

    public Card(final Context context, AttributeSet attrs) {
        super(context, attrs);

        final LayoutInflater inflater = (LayoutInflater) context
                .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        inflater.inflate(R.layout.card, this);

        if (!isInEditMode()) {
            preferenceManager = new MyPreferenceManager(context);
        }

        label = (TextView) findViewById(R.id.card_title);
        info = (TextView) findViewById(R.id.card_info);
        entry = (EditText) findViewById(R.id.card_entry);

        entry.addTextChangedListener(new MyTextWatcher());
        entry.setOnFocusChangeListener(new MyOnFocusChangeListener());
        entry.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                if (actionId == EditorInfo.IME_ACTION_DONE) {
                    entry.clearFocus();

                    InputMethodManager inputMethodManager = (InputMethodManager) context
                            .getSystemService(
                                    Context.INPUT_METHOD_SERVICE);
                    inputMethodManager.hideSoftInputFromWindow(entry.getWindowToken(), 0);
                }
                return false;
            }
        });

        this.setOnClickListener(new MyOnClickListener());

        TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.Card, 0, 0);

        try {
            prefKey = typedArray.getString(R.styleable.Card_prefKey);
            label.setText(typedArray.getString(R.styleable.Card_labelText));
            info.setText(typedArray.getString(R.styleable.Card_infoText));
            entry.setHint(typedArray.getString(R.styleable.Card_hintText));
        } finally {
            typedArray.recycle();
        }

        // Changes the hint text of cards related to glucose level based on the preference chose
        boolean isCardGlucoseLevel = (getId() == R.id.card_desired_blood_glucose_level ||
                getId() == R.id.card_corrective_factor ||
                getId() == R.id.card_current_blood_glucose_level);

        if (isCardGlucoseLevel) {
            if (!isInEditMode()) {
                switch (preferenceManager.getBloodGlucoseUnit()) {
                    case mmol:
                        entry.setHint(getResources().getString(R.string.hint_mmol));
                        break;
                    case mgdl:
                        entry.setHint(getResources().getString(R.string.hint_mgdl));
                }
            } else {
                entry.setHint(getResources().getString(R.string.hint_mmol));
            }
        }
    }

    /**
     * Adds a point to one decimal place of a string of characters.
     */
    public static void addDecimalPlace(CharSequence s, EditText et) {
        List<Integer> storedPoints = new ArrayList<>();

        if (!textChangeRunning) {
            textChangeRunning = true;

            StringBuilder decimalPlaceAuto = new StringBuilder(s.toString());

            while (decimalPlaceAuto.length() > 2 && decimalPlaceAuto
                    .charAt(0) == '0' || decimalPlaceAuto.charAt(0) == '.') {
                decimalPlaceAuto.deleteCharAt(0);
            }

            for (int i = 0; i < decimalPlaceAuto.length(); i++) {
                if (decimalPlaceAuto.charAt(i) == '.') {
                    storedPoints.add(i);
                }
            }

            for (int i = 0; i < storedPoints.size(); i++) {
                decimalPlaceAuto.deleteCharAt(storedPoints.get(i));
            }

            while (decimalPlaceAuto.length() < 2) {
                decimalPlaceAuto.insert(0, '0');
            }

            decimalPlaceAuto.insert(decimalPlaceAuto.length() - 1, '.');

            et.setText(decimalPlaceAuto.toString());

            textChangeRunning = false;

            Selection.setSelection(et.getText(), decimalPlaceAuto.toString().length());

        }
    }

    // Methods
    public void resetEntryField() {
        entry.setText("");
    }

    public double getValueFromEntryField() {
        String string = entry.getText().toString();

        if (string.isEmpty()) {
            return 0;
        } else {
            return Double.valueOf(string);
        }
    }

    public void setEntryFieldFromValue(Double value) {
        if (shouldDecimalPlaceBeDisplayed(getId())) {
            entry.setText(String.valueOf(value));
        } else {
            long integer = Math.round(value);
            entry.setText(String.valueOf(integer));
        }
    }

    public boolean isEntryFieldFilled() {
        return (entry.getText().toString().length() > 0);
    }

    private boolean shouldDecimalPlaceBeDisplayed(int cardId) {
        boolean isCardCarbohydratesInMeal = (cardId == R.id.card_carbohydrates_in_meal);

        boolean isCardGlucoseLevel = (cardId == R.id.card_desired_blood_glucose_level ||
                getId() == R.id.card_corrective_factor ||
                getId() == R.id.card_current_blood_glucose_level);

        boolean isTotalDailyDoseCard = (cardId == R.id.card_total_daily_dose);

        if (isCardCarbohydratesInMeal) {
            if (!isInEditMode()) {
                return preferenceManager.allowFloatingPointCarbohydrates();
            } else {
                return false;
            }
        } else if (isCardGlucoseLevel) {
            if (!isInEditMode()) {
                switch (preferenceManager.getBloodGlucoseUnit()) {
                    case mmol:
                        return true;
                    case mgdl:
                        return false;
                }
            } else {
                return true;
            }
        } else if (isTotalDailyDoseCard) {
            return false;
        }

        return true;
    }

    public void setOnTextChangeListener(OnTextChangeListener textChangeListener) {
        this.textChangeListener = textChangeListener;
    }

    public interface OnTextChangeListener {
        void onTextChange();
    }

    public class MyOnClickListener implements OnClickListener {

        @Override
        public void onClick(View v) {
            entry.requestFocus();
            entry.selectAll();
        }
    }

    private class MyTextWatcher implements TextWatcher {
        @Override
        public void beforeTextChanged(CharSequence charSequence, int i, int i2, int i3) {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {
            modifyText(s, shouldDecimalPlaceBeDisplayed(getId()));

            if (textChangeListener != null) {
                textChangeListener.onTextChange();
            }
        }

        @Override
        public void afterTextChanged(Editable editable) {

        }

        private void modifyText(CharSequence s, boolean addDecimal) {
            if (entry.length() > 0) {
                if (addDecimal) {
                    addDecimalPlace(s, entry);
                }

                entry.setTextSize(TypedValue.COMPLEX_UNIT_PX,
                        getResources().getDimension(R.dimen.text_card_entry_filled));
            } else {
                entry.setTextSize(TypedValue.COMPLEX_UNIT_PX,
                        getResources().getDimension(R.dimen.text_card_entry_empty));
            }
        }
    }

    private class MyOnFocusChangeListener implements OnFocusChangeListener {

        @Override
        public void onFocusChange(View v, boolean hasFocus) {
            if (!hasFocus) {
                double value = getValueFromEntryField();

                if (value != 0 && prefKey != null) {
                    preferenceManager.editor.putFloat(prefKey, (float) value).commit();
                }
            }
        }
    }
}

File: Insulator/app/src/main/java/com/bendywalker/insulator/Card.kt
package com.bendywalker.insulator

import android.content.Context
import android.text.*
import android.util.AttributeSet
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.TextView

/**
 * Created by Ben David Walker (bendywalker) on 11/01/2017.
 */
class Card(context: Context, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) : LinearLayout(context, attrs, defStyleAttr, defStyleRes) {
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : this(context, attrs, defStyleAttr, 0)
    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0, 0)
    constructor(context: Context) : this(context, null, 0, 0)

    private val titleTextView by lazy { findViewById(R.id.textView_card_title) as TextView }
    private val descriptionTextView by lazy { findViewById(R.id.textView_card_description) as TextView }
    private val entryEditText by lazy { findViewById(R.id.editText_card_entry) as EditText }
    private val persistedValues by lazy { PersistedValues(context) }

    private val persistedValueKey: String?
    private var addingFloatingPoint = false

    var onTextChangeListener: OnTextChangeListener? = null

    var value: Double
        get() {
            val string = entryEditText.text.toString()
            return if (string.isEmpty()) 0.0 else string.toDouble()
        }
        set(value) {
            entryEditText.setText(value.toString())
        }

    private val shouldDisplayFloatingPoint: Boolean
        get() {
            // TODO: Write logic for booleans once other layouts have been implemented
            val carbohydratesInMeal = id == R.id.card_variableData_carbohydratesInMeal
            val glucoseLevel = id == R.id.card_variableData_currentBloodGlucoseLevel
            val totalDailyDose = true

            if (carbohydratesInMeal) {
                return persistedValues.allowFloatingPointCarbohydrates
            } else if (glucoseLevel) {
                when (persistedValues.bloodGlucoseUnit) {
                    BloodGlucoseUnit.MMOL -> return true
                    BloodGlucoseUnit.MGDL -> return false
                }
            } else if (totalDailyDose) {
                return false
            }

            return true
        }

    init {
        val layoutInflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        layoutInflater.inflate(R.layout.view_card, this)

        val typedArray = context.obtainStyledAttributes(attrs, R.styleable.Card, 0, 0)
        try {
            persistedValueKey = typedArray.getString(R.styleable.Card_persistedValueKey)
            titleTextView.text = typedArray.getString(R.styleable.Card_titleText)
            descriptionTextView.text = typedArray.getString(R.styleable.Card_descriptionText)
            entryEditText.hint = typedArray.getString(R.styleable.Card_hintText)
        } finally {
            typedArray.recycle()
        }

        if (isInEditMode) {
            // Do this to have standardised values in the layout preview
            entryEditText.setText(R.string.dose_placeholder)
            entryEditText.adjustTextSize()
            entryEditText.setMaxLength(4)
        } else {
            setOnClickListener {
                entryEditText.requestFocus()
                entryEditText.selectAll()
            }

            entryEditText.setOnTextChangedListener { s ->
                if (!TextUtils.isEmpty(entryEditText.text)) {
                    if (shouldDisplayFloatingPoint && !addingFloatingPoint) entryEditText.addFloatingPoint()
                }
                entryEditText.adjustTextSize()
                onTextChangeListener?.onTextChange()
            }

            entryEditText.setOnFocusChangeListener { view, hasFocus -> if (!hasFocus) persistedValues.setValueForKey(value, persistedValueKey) }

            entryEditText.setOnEditorActionListener { textView, actionId, keyEvent ->
                if (actionId == EditorInfo.IME_ACTION_DONE) {
                    entryEditText.clearFocus()
                    val inputMethodManager = context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                    inputMethodManager.hideSoftInputFromWindow(entryEditText.windowToken, 0)
                }

                false
            }

            entryEditText.setMaxLength(if (shouldDisplayFloatingPoint) 4 else 3)
        }
    }

    fun reset() {
        entryEditText.setText("")
    }

    fun EditText.setOnTextChangedListener(closure: (s: CharSequence) -> Unit) {
        this.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable?) {
                closure(s as CharSequence)
            }
        })
    }

    fun EditText.addFloatingPoint() {
        addingFloatingPoint = true
        val string = addFloatingPoint(this.text.toString())
        this.setText(string)
        addingFloatingPoint = false
        Selection.setSelection(this.text, string.length)
    }

    fun EditText.adjustTextSize() {
        if (TextUtils.isEmpty(this.text))
            this.setTextSize(TypedValue.COMPLEX_UNIT_PX, resources.getDimension(R.dimen.text_card_entry_empty))
        else this.setTextSize(TypedValue.COMPLEX_UNIT_PX, resources.getDimension(R.dimen.text_card_entry_filled))
    }

    fun EditText.setMaxLength(length: Int) {
        this.filters = listOf<InputFilter>(InputFilter.LengthFilter(length)).toTypedArray()
    }

    companion object {
        fun addFloatingPoint(string: String, precision: Int = 1): String {
            val stringBuilder = StringBuilder(string)
            while (stringBuilder.length > 2 && stringBuilder.first() == '0' || stringBuilder.first() == '.') stringBuilder.deleteCharAt(0)
            val storedPoints = (0 until stringBuilder.length).filter { stringBuilder[it] == '.' }
            for (i in storedPoints) stringBuilder.deleteCharAt(i)
            while (stringBuilder.length < 2) stringBuilder.insert(0, '0')
            stringBuilder.insert(stringBuilder.length - precision, '.')
            return stringBuilder.toString()
        }
    }

    interface OnTextChangeListener {
        fun onTextChange()
    }
}